src/dual.jl

```diff
@@ -384,13 +384,35 @@ for pred in UNARY_PREDICATES
     @eval Base.$(pred)(d::Dual) = $(pred)(value(d))
 end
 
-for pred in BINARY_PREDICATES
+# BINARY_PREDICATES = Symbol[:isequal, :isless, :<, :>, :(==), :(!=), :(<=), :(>=)] # prelude.jl
+
+Base.iszero(x::Dual) = iszero(value(x)) && iszero(partials(x))
+
+for pred in [:isequal, :(==)]
+    @eval begin
+        @define_binary_dual_op(
+            Base.$(pred),
+            $(pred)(value(x), value(y)) && $(pred)(partials(x), partials(y)),
+            $(pred)(value(x), y)        && iszero(partials(x)),
+            $(pred)(x, value(y))        && iszero(partials(y)),
+        )
+    end
+end
+
+@define_binary_dual_op(
+    Base.:(!=),
+    (!=)(value(x), value(y)) || (!=)(partials(x), partials(y)),
+    (!=)(value(x), y)        || !iszero(partials(x)),
+    (!=)(x, value(y))        || !iszero(partials(y)),
+)
```

Comment:
Is this needed? Isn't this covered by the default definition of `!=` (https://github.com/JuliaLang/julia/blob/742b9abb4dd4621b667ec5bb3434b8b3602f96fd/base/operators.jl#L278) and the definition of `==` above? If it's needed (ambiguity issues?), then it might also be helpful to add a short comment that this just ensures that `!=(x, y) = !(x == y)`.

Reply:
The original code defines all of `[:isequal, :isless, :<, :>, :(==), :(!=), :(<=), :(>=)]`, even though some could call others via Base's definitions. Why, I'm not sure, maybe there were efficiency or ambiguity reasons. I guess I decided not to change that, to touch fewer things.

Reply:
Yes, I assumed there might be a reason. Maybe just add a short comment explaining the reasoning behind this definition (i.e., it's just `!=(x, y) = !(x == y)`)?